# Projects Overviews 

# NPL Bowling Analytics and Performance Analysis System

## Project Overview

This project is a comprehensive bowling analytics system developed using SQL and Python (Pandas) to analyze Nepal Premier League (NPL) ball-by-ball data. The objective of the project is to transform raw cricket data into meaningful insights that help evaluate bowler performance across different match situations.

The project replicates advanced Indian Premier League (IPL) analytics and applies the same statistical methods to NPL data.

---

## Purpose of the Project

Modern cricket increasingly depends on data-driven decision-making. Traditional statistics such as total wickets or total runs conceded often fail to explain a player's true impact.

The purpose of this project is to:

* Measure bowler effectiveness more accurately.
* Evaluate performance in different match phases.
* Identify strengths and weaknesses of bowlers.
* Support team selection and strategic decisions.
* Provide analytical insights similar to professional T20 leagues.

---

## Major Analyses Performed

### Overall Bowling Performance

* Total wickets
* Runs conceded
* Economy rate
* Matches played
* Best bowling figures

### Match Performance Analysis

* Best bowling figures in a match
* Five-wicket hauls
* Most wickets in a match

### Economy Analysis

* Best economy bowlers
* Worst economy bowlers

### Match Phase Analysis

* Powerplay bowlers (Overs 1–6)
* Middle-over bowlers (Overs 7–15)
* Death-over bowlers (Overs 16–20)

### Over Analysis

* Most expensive overs
* Runs conceded in a single over
* Boundary analysis (fours and sixes)

### Wicket Analysis

* Highest wicket takers
* Dismissal type analysis
* Wicket rate analysis

---

## Technologies Used

* MySQL
* Python
* Pandas
* Matplotlib
* Jupyter Notebook

---

## Business Value

This system can assist:

* Team management
* Coaches
* Selectors
* Performance analysts
* Cricket franchises

The insights generated help in:

* Selecting the best bowlers.
* Planning bowling strategies.
* Identifying death-over specialists.
* Improving team combinations.
* Opponent analysis.
* Player scouting and recruitment.

---

## Advantages of the System

### Data-Driven Decision Making

Decisions are based on statistics rather than assumptions.

### Performance Evaluation

Provides objective measurement of player performance.

### Role Identification

Identifies:

* Powerplay specialists
* Middle-over controllers
* Death-over specialists

### Talent Identification

Helps discover emerging bowlers based on performance metrics.

### Strategy Development

Supports match planning and bowling rotations.

### Performance Monitoring

Tracks player consistency across seasons.

---

## Conclusion

This project demonstrates how data analytics can be applied to cricket to generate actionable insights. By converting raw ball-by-ball data into meaningful performance metrics, the system enables evidence-based evaluation of bowlers and supports strategic decision-making in modern cricket.


# Import The Library 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Read the datasets 

In [2]:
_2024=pd.read_csv("NPL-2024.csv")
_2025=pd.read_csv("NPL-2025.csv")

# Data Preprocessing Steps 

In [6]:
_2025.columns

Index(['match_no', 'inning', 'batting_team', 'bowling_team', 'ball_over',
       'ball_result', 'bowler', 'batter', 'non_striker', 'shot_direction',
       'player_out', 'dismissal_kind', 'fielders', 'batsman_runs', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'extra_runs', 'total_runs'],
      dtype='object')

In [7]:
_2024.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'ball_over',
       'ball_result', 'bowler', 'batsman', 'non_striker', 'shot_direction',
       'player_dismissed', 'dismissal_kind', 'fielder', 'batsman_runs',
       'wide_runs', 'bye_runs', 'legbye_runs', 'noball_runs', 'extra_runs',
       'total_runs'],
      dtype='object')

In [10]:
_2025.rename(columns={"match_no":"match_id","batter":"batsman","player_out":"player_dismissed"},inplace=True)

# Combine the datasets 

In [18]:
final=pd.concat([_2024,_2025],keys=["2024","2025"],ignore_index=False)
pd.set_option("display.max_columns",None)
final=final.reset_index()
final=final.drop(columns=["level_1"]).rename(columns={"level_0":"season"})

In [27]:
final

,season,match_id,inning,batting_team,bowling_team,ball_over,ball_result,bowler,batsman,non_striker,shot_direction,player_dismissed,dismissal_kind,fielder,batsman_runs,wide_runs,bye_runs,legbye_runs,noball_runs,extra_runs,total_runs,fielders
0,2024,1,1,Biratnagar Kings,Janakpur Bolts,0.1,0,Kishore Mahato,Lokesh Bam,Martin Guptill,NaN,NaN,NaN,NaN,0,0,0,0,0,0,0,NaN
1,2024,1,1,Biratnagar Kings,Janakpur Bolts,0.2,4,Kishore Mahato,Lokesh Bam,Martin Guptill,third man,NaN,NaN,NaN,4,0,0,0,0,0,4,NaN
2,2024,1,1,Biratnagar Kings,Janakpur Bolts,0.3,0,Kishore Mahato,Lokesh Bam,Martin Guptill,NaN,NaN,NaN,NaN,0,0,0,0,0,0,0,NaN
3,2024,1,1,Biratnagar Kings,Janakpur Bolts,0.4,1w,Kishore Mahato,Lokesh Bam,Martin Guptill,NaN,NaN,NaN,NaN,0,1,0,0,0,1,1,NaN
4,2024,1,1,Biratnagar Kings,Janakpur Bolts,0.4,0,Kishore Mahato,Lokesh Bam,Martin Guptill,NaN,NaN,NaN,NaN,0,0,0,0,0,0,0,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14919,2025,32,2,Lumbini Lions,Sudurpaschim Royals,8.2,0,Dipendra Singh Airee,Niroshan Dickwella,Sundeep Jora,covers,NaN,NaN,NaN,0,0,0,0,0,0,0,NaN
14920,2025,32,2,Lumbini Lions,Sudurpaschim Royals,8.3,1,Dipendra Singh Airee,Niroshan Dickwella,Sundeep Jora,point,NaN,NaN,NaN,1,0,0,0,0,0,1,NaN
14921,2025,32,2,Lumbini Lions,Sudurpaschim Royals,8.4,0,Dipendra Singh Airee,Sundeep Jora,Niroshan Dickwella,square leg,NaN,NaN,NaN,0,0,0,0,0,0,0,NaN
14922,2025,32,2,Lumbini Lions,Sudurpaschim Royals,8.5,W,Dipendra Singh Airee,Sundeep Jora,Niroshan Dickwella,covers,Sundeep Jora,caught,NaN,0,0,0,0,0,0,0,Abinash Bohara


In [88]:
final.sample(5)

,season,match_id,inning,batting_team,bowling_team,ball_over,ball_result,bowler,batsman,non_striker,shot_direction,player_dismissed,dismissal_kind,fielder,batsman_runs,wide_runs,bye_runs,legbye_runs,noball_runs,extra_runs,total_runs,fielders,over
1197,2024,6,1,Karnali Yaks,Kathmandu Gurkhas,13.5,W,Gerhard Erasmus,Gulshan Jha,Shikhar Dhawan,NaN,Gulshan Jha,stumped,Stephen Eskinazi,0,0,0,0,0,0,0,NaN,13
1182,2024,6,1,Karnali Yaks,Kathmandu Gurkhas,11.5,1,Nathan Sowter,Shikhar Dhawan,Gulshan Jha,square leg,NaN,NaN,NaN,1,0,0,0,0,0,1,NaN,11
5459,2024,24,1,Pokhara Avengers,Kathmandu Gurkhas,16.3,1,Karan KC,Bipin Khatri,Raymon Reifer,fine leg,NaN,NaN,NaN,1,0,0,0,0,0,1,NaN,16
4994,2024,22,2,Lumbini Lions,Karnali Yaks,1.3,4,Nandan Yadav,Aashutosh Ghiraiya,Arjun Saud,covers,NaN,NaN,NaN,4,0,0,0,0,0,4,NaN,1
12534,2025,22,1,Biratnagar Kings,Lumbini Lions,15.6,1,Sher Malla,Pratish GC,Basir Ahamad,mid wicket,NaN,NaN,NaN,1,0,0,0,0,0,1,NaN,15


# Question 1 : Bowlers who bollwed most of the deliverires 

## sql equivalents 

In [ ]:
SELECT
    bowler,
    COUNT(*) AS deliveries
FROM npl_ball_by_ball
GROUP BY bowler
ORDER BY deliveries DESC;

## pandas equivalents 

In [30]:
bowler_deliveries = (
    final.groupby("bowler")
      .size()
      .reset_index(name="deliveries")
      .sort_values("deliveries", ascending=False)
)

bowler_deliveries.head(10)


,bowler,deliveries
109,Sandeep Lamichhane,389
63,Lalit Rajbanshi,360
114,Scott Kuggeleijn,353
121,Sompal Kami,341
118,Sher Malla,336
57,Karan KC,331
46,Harmeet Singh,326
115,Shahab Alam,308
80,Nandan Yadav,307
120,Sohail Tanvir,307


### Conclusion: Bowlers with the Most Deliveries in NPL

The analysis of bowlers who have delivered the highest number of balls in the Nepal Premier League highlights the bowlers that teams rely on most heavily throughout the tournament. These players have consistently been entrusted with bowling responsibilities across multiple matches and critical situations.

Sandeep Lamichhane leads the list with 389 deliveries, indicating his role as a primary strike bowler and a key component of his team's bowling attack. Lalit Rajbanshi, Scott Kuggeleijn, Sompal Kami, and Sher Malla also appear among the most utilized bowlers, demonstrating their consistency, fitness, and importance to their respective teams.

A high number of deliveries generally reflects several important factors:

* Team trust and confidence in the bowler.
* Ability to bowl in multiple phases of the innings.
* Consistency in team selection.
* Physical fitness and workload management.
* Experience and reliability under pressure.

This analysis provides valuable insights into bowling workload distribution and helps identify the core bowlers around whom teams build their bowling strategies. Bowlers with higher delivery counts often have a greater impact on match outcomes and can be further evaluated using metrics such as economy rate, wicket-taking ability, and phase-wise performance.

Overall, the delivery analysis serves as a foundation for assessing bowler consistency, durability, and their strategic importance within the Nepal Premier League.


In [33]:
final["dismissal_kind"].unique()

array([nan, 'Bowled', 'LBW', 'Caught', 'Runout', 'Stumped',
       'Caught & Bowled', 'Retired Hurt', 'caught', 'bowled', 'stumped',
       'caught & bowled', 'lbw', 'run out', 'hit wicket', 'retired hurt'],
      dtype=object)

# Question 2 : Highest Wicket takers in NPL

In [36]:
final["dismissal_kind"] = (
    final["dismissal_kind"]
    .str.lower()
    .str.strip()
)

# sql equivalents 

In [ ]:
SELECT
    bowler,
    COUNT(*) AS wickets,
    SUM(CASE WHEN dismissal_kind IN ('caught') THEN 1 ELSE 0 END) AS caught,
    SUM(CASE WHEN dismissal_kind IN ('bowled') THEN 1 ELSE 0 END) AS bowled,
    SUM(CASE WHEN dismissal_kind IN ('lbw') THEN 1 ELSE 0 END) AS lbw,
    SUM(CASE WHEN dismissal_kind IN ('stumped') THEN 1 ELSE 0 END) AS stumped,
    SUM(CASE WHEN dismissal_kind IN ('caught & bowled') THEN 1 ELSE 0 END) AS caught_bowled,
    SUM(CASE WHEN dismissal_kind IN ('hit wicket') THEN 1 ELSE 0 END) AS hit_wicket
FROM npl_ball_by_ball

WHERE dismissal_kind IS NOT NULL
AND dismissal_kind NOT IN (
    'runout',
    'run out',
    'retired hurt'
)
GROUP BY bowler
ORDER BY wickets DESC;

### pandas equivalents 

In [37]:
valid_wickets = final[
    (~final["dismissal_kind"].isin(
        ["runout", "run out", "retired hurt"]
    )) &
    (final["dismissal_kind"].notna())
]

result = (
    valid_wickets
    .groupby("bowler")
    .agg(
        wickets=("dismissal_kind", "count"),
        caught=("dismissal_kind",lambda x: (x == "caught").sum()),
        bowled=("dismissal_kind",lambda x: (x == "bowled").sum()),
        lbw=("dismissal_kind",lambda x: (x == "lbw").sum()),
        stumped=("dismissal_kind",lambda x: (x == "stumped").sum()),
        caught_bowled=("dismissal_kind",lambda x: (x == "caught & bowled").sum()),
        hit_wicket=("dismissal_kind",lambda x: (x == "hit wicket").sum())
    )
    .reset_index()
    .sort_values("wickets", ascending=False)
)

result.head(10)

,bowler,wickets,caught,bowled,lbw,stumped,caught_bowled,hit_wicket
88,Scott Kuggeleijn,27,14,8,4,0,1,0
42,Karan KC,25,17,4,3,0,1,0
91,Sher Malla,24,12,7,4,1,0,0
4,Abinash Bohara,24,19,4,1,0,0,0
84,Sandeep Lamichhane,23,16,4,2,0,1,0
93,Sohail Tanvir,22,15,2,3,1,1,0
47,Lalit Rajbanshi,21,13,5,1,2,0,0
32,Harmeet Singh,20,10,2,4,3,1,0
61,Nandan Yadav,19,11,7,1,0,0,0
89,Shahab Alam,19,11,2,3,3,0,0


### Conclusion: Highest Wicket-Taking Bowlers in NPL

The wicket analysis reveals the most successful wicket-taking bowlers in the Nepal Premier League and provides insights into their bowling styles and methods of dismissal.

Scott Kuggeleijn emerged as the leading wicket-taker with 27 wickets, demonstrating his ability to consistently break partnerships and provide crucial breakthroughs. Karan KC and Sher Malla closely followed with 25 and 24 wickets respectively, highlighting their importance as strike bowlers.

A significant observation from the analysis is that the majority of dismissals occurred through catches. Bowlers such as Abinash Bohara and Karan KC recorded a high proportion of caught dismissals, indicating their effectiveness in generating edges, inducing mistimed shots, and utilizing field placements effectively.

Several bowlers also demonstrated strong wicket-taking abilities through direct bowling dismissals. Scott Kuggeleijn and Sher Malla recorded high numbers of bowled dismissals, suggesting their ability to beat batters with pace, movement, and accuracy.

Spin bowlers such as Lalit Rajbanshi, Harmeet Singh, and Shahab Alam contributed through LBW and stumping dismissals, indicating their effectiveness in deceiving batters through flight, turn, and variations. The presence of stumping dismissals further reflects the impact of spin bowling during the middle overs.

The analysis also indicates that caught dismissals remain the dominant mode of wicket-taking in T20 cricket, emphasizing the importance of disciplined bowling and effective field placements.

Overall, this analysis helps identify:

* Strike bowlers who regularly take wickets.
* Bowlers who create chances through catches.
* Bowlers capable of dismissing batters directly through bowled and LBW dismissals.
* Spin bowlers who rely on deception and turn.
* Fast bowlers who generate wickets through pace and movement.

These insights can support team selection, bowling strategy, player development, and opposition analysis while helping teams understand the strengths and wicket-taking patterns of their bowlers.


# Question 3 :  Highest wicket Talen By the bowler in an NPL matches

### sql queries 

In [ ]:
SELECT
    match_id,
    bowler,
    COUNT(*) AS wickets,
    SUM(total_runs) AS runs_conceded,
    CONCAT(
        COUNT(*),
        '-',
        SUM(total_runs)
    ) AS best_figures

FROM npl_ball_by_ball
WHERE dismissal_kind IS NOT NULL
AND LOWER(dismissal_kind) NOT IN (
    'runout',
    'run out',
    'retired hurt'
)

GROUP BY match_id, bowler
ORDER BY wickets DESC,
         runs_conceded ASC;

## pandas equivalents 

In [39]:
valid = final[
    (final["dismissal_kind"].notna()) &
    (~final["dismissal_kind"].isin(
        ["runout", "run out", "retired hurt"]
    ))
]

wickets = (
    valid.groupby(["match_id", "bowler"])
    .size()
    .reset_index(name="wickets")
)

runs = (
    final.groupby(["match_id", "bowler"])["total_runs"]
      .sum()
      .reset_index(name="runs")
)

result = wickets.merge(
    runs,
    on=["match_id", "bowler"]
)

result["best_figures"] = (
    result["wickets"].astype(str)
    + "-"
    + result["runs"].astype(str)
)

result = result.sort_values(
    ["wickets", "runs"],
    ascending=[False, True]
)

result.head()

,match_id,bowler,wickets,runs,best_figures
333,23,Shahab Alam,6,25,6-25
247,17,William Bosisto,6,28,6-28
376,26,Sandeep Lamichhane,6,44,6-44
24,2,Sohail Tanvir,6,51,6-51
389,27,Scott Kuggeleijn,5,18,5-18


### Conclusion: Best Bowling Figures in an NPL Match

The analysis of the best bowling figures in a single match highlights the most dominant bowling performances in the Nepal Premier League. These performances demonstrate a bowler's ability to control the game, take multiple wickets, and significantly influence match outcomes.

Shahab Alam recorded the best bowling performance with exceptional figures of **6 wickets for 25 runs**, making it the most effective bowling spell among the analyzed matches. This performance reflects outstanding control, wicket-taking ability, and match-winning impact.

William Bosisto also delivered an impressive spell of **6 wickets for 28 runs**, showing remarkable effectiveness in restricting the opposition while consistently taking wickets. Similarly, Sandeep Lamichhane achieved **6 wickets for 44 runs**, proving his ability to take wickets even while bowling attacking spells.

Sohail Tanvir's figures of **6 wickets for 51 runs** indicate an aggressive wicket-taking approach, where the bowler continued attacking despite conceding additional runs. Meanwhile, Scott Kuggeleijn's outstanding figures of **5 wickets for 18 runs** represent one of the most economical and impactful bowling performances, combining both wicket-taking ability and run control.

Several important observations can be drawn from these performances:

* Multiple six-wicket hauls indicate the presence of genuine match-winning bowlers in the tournament.
* Lower runs conceded alongside high wicket counts demonstrate superior bowling efficiency.
* Exceptional bowling figures often result in significant momentum shifts and directly influence match results.
* Bowlers capable of taking five or more wickets in a match become valuable assets for their teams.

Overall, these bowling performances represent some of the finest individual spells in the Nepal Premier League. They highlight bowlers who possess the ability to dominate batting lineups, change the course of matches, and deliver under pressure. Such performances are critical indicators of bowling excellence and can assist teams in identifying their most impactful and dependable bowlers.


#  Question No 4 : No of 5-wicket hauls by bowlers in an NPL

In [ ]:
WITH match_wickets AS (
    SELECT
        match_id,
        bowler,
        COUNT(*) AS wickets
    FROM npl_ball_by_ball
    WHERE dismissal_kind IS NOT NULL
    AND LOWER(dismissal_kind) NOT IN (
        'runout',
        'run out',
        'retired hurt'
    )
    GROUP BY match_id, bowler
)

SELECT
    bowler,
    COUNT(*) AS five_wicket_hauls
FROM match_wickets
WHERE wickets >= 5
GROUP BY bowler
ORDER BY five_wicket_hauls DESC;

In [41]:
final["dismissal_kind"] = (
    final["dismissal_kind"]
      .str.lower()
      .str.strip()
)

valid_wickets = final[
    (final["dismissal_kind"].notna()) &
    (~final["dismissal_kind"].isin(
        ["runout", "run out", "retired hurt"]
    ))
]

match_wickets = (
    valid_wickets
    .groupby(["match_id", "bowler"])
    .size()
    .reset_index(name="wickets")
)

five_wicket_hauls = (
    match_wickets[
        match_wickets["wickets"] >= 5
    ]
    .groupby("bowler")
    .size()
    .reset_index(name="hauls")
    .sort_values("hauls", ascending=False)
)

five_wicket_hauls

,bowler,hauls
6,Scott Kuggeleijn,2
8,Sohail Tanvir,2
1,Karan KC,1
0,Harsh Thaker,1
2,Naren Saud,1
3,Ranjeet Kumar,1
5,Sandeep Lamichhane,1
4,Rijan Dhakal,1
7,Shahab Alam,1
9,William Bosisto,1


### Conclusion: Five-Wicket Hauls in the Nepal Premier League

The analysis of five-wicket hauls highlights the bowlers who have produced exceptional match-winning performances in the Nepal Premier League. Taking five or more wickets in a T20 match is considered a significant achievement, as the limited number of overs makes it difficult for bowlers to dismiss multiple batters in a single innings.

Scott Kuggeleijn and Sohail Tanvir emerged as the most successful bowlers in this category, each recording **two five-wicket hauls**. Their ability to repeatedly produce outstanding performances demonstrates consistency, experience, and the capability to dominate opposition batting lineups.

Several other bowlers, including Karan KC, Harsh Thaker, Naren Saud, Ranjeet Kumar, Sandeep Lamichhane, Rijan Dhakal, Shahab Alam, and William Bosisto, each recorded one five-wicket haul. These performances indicate that a wide range of bowlers have been capable of delivering match-winning spells under different match situations.

The analysis provides several important insights:

* Multiple five-wicket hauls indicate exceptional wicket-taking ability and consistency.
* Bowlers who achieve these milestones often play a decisive role in determining match outcomes.
* Five-wicket performances demonstrate a bowler's ability to maintain pressure and exploit batting weaknesses.
* Such performances often occur during crucial stages of the match and can significantly shift momentum.

Overall, the five-wicket haul analysis identifies the elite wicket-taking bowlers of the Nepal Premier League. These bowlers have demonstrated the ability to deliver exceptional individual performances, making them valuable assets for their teams and key contributors to match victories.


# Question No 5 : Most Runs Conceded by a Bowler in an NPL Match

In [ ]:
SELECT
    match_id,
    bowler,
    SUM(total_runs) AS runs_conceded
FROM npl_ball_by_ball
GROUP BY
    match_id,
    bowler
ORDER BY runs_conceded DESC;

In [42]:
bowler_runs = (
    final.groupby(["match_id", "bowler"])["total_runs"]
      .sum()
      .reset_index(name="runs_conceded")
      .sort_values("runs_conceded", ascending=False)
)

bowler_runs.head(10)

,match_id,bowler,runs_conceded
743,32,Dipendra Singh Airee,69
757,32,Scott Kuggeleijn,66
139,7,Bibek Yadav,64
411,18,Lalit Rajbanshi,64
231,10,Sompal Kami,64
38,2,Rijan Dhakal,61
432,19,Rashid Khan,60
51,3,James Neesham,59
321,14,Lalit Rajbanshi,57
404,18,Dipendra Singh Airee,57


### Conclusion: Most Runs Conceded by a Bowler in an NPL Match

The analysis of the highest runs conceded in a single match identifies the most expensive bowling performances in the Nepal Premier League. While these figures represent challenging outings for bowlers, they also provide valuable insights into match conditions, batting dominance, and the risks associated with aggressive bowling strategies in T20 cricket.

Dipendra Singh Airee recorded the highest runs conceded in a single match, giving away **69 runs**, followed by Scott Kuggeleijn with **66 runs**. Bibek Yadav, Lalit Rajbanshi, and Sompal Kami also experienced difficult bowling performances, conceding more than 60 runs in a match.

Several important observations can be drawn from these results:

* High-scoring T20 matches often place significant pressure on bowlers, especially during the death overs.
* Bowlers who regularly bowl during powerplay or death overs are generally exposed to greater scoring opportunities and therefore may concede more runs.
* Some bowlers appearing in this list are also among the leading wicket-takers or most-used bowlers, indicating that captains continue to trust them in crucial situations despite occasional expensive spells.
* Aggressive wicket-taking bowlers often accept the risk of conceding additional runs while attempting to dismiss batters.

The presence of experienced bowlers such as Dipendra Singh Airee, Rashid Khan, James Neesham, and Scott Kuggeleijn in this list demonstrates that even high-quality bowlers can have difficult matches in the T20 format. Such performances should therefore be evaluated alongside other performance metrics such as wickets, economy rate, bowling phases, and overall consistency.

Overall, this analysis highlights the unpredictable nature of T20 cricket and emphasizes the importance of evaluating bowlers using long-term performance indicators rather than a single expensive spell. These insights can help teams understand bowling risks, workload management, and tactical decisions during high-scoring matches.


 # Question No 6 : Highest Runs Concede in an NPL over by a bowler

In [ ]:
WITH over_data AS (
    SELECT
        match_id,
        innings,
        FLOOR(over_ball) AS over_no,
        bowler,

        SUM(total_runs) AS runs,

        SUM(
            CASE
                WHEN batsman_runs = 4 THEN 1
                ELSE 0
            END
        ) AS fours,

        SUM(
            CASE
                WHEN batsman_runs = 6 THEN 1
                ELSE 0
            END
        ) AS sixes,

        SUM(wide) AS wides,
        SUM(noball) AS noballs,
        SUM(legbye) AS legbyes,
        SUM(bye) AS byes

    FROM npl_ball_by_ball

    GROUP BY
        match_id,
        innings,
        FLOOR(over_ball),
        bowler
)

SELECT
    bowler,
    match_id,
    innings,
    over_no,
    runs,
    fours,
    sixes,
    wides,
    noballs,
    legbyes,
    byes

FROM over_data

ORDER BY runs DESC;

In [46]:
final["over"] = final["ball_over"].astype(str).str.split(".").str[0].astype(int)

In [45]:
final.columns

Index(['season', 'match_id', 'inning', 'batting_team', 'bowling_team',
       'ball_over', 'ball_result', 'bowler', 'batsman', 'non_striker',
       'shot_direction', 'player_dismissed', 'dismissal_kind', 'fielder',
       'batsman_runs', 'wide_runs', 'bye_runs', 'legbye_runs', 'noball_runs',
       'extra_runs', 'total_runs', 'fielders'],
      dtype='object')

In [49]:
over_analysis = (
    final.groupby(
        ["match_id", "inning", "over", "bowler"]
    )
    .agg(
        runs=("total_runs", "sum"),
        fours=("batsman_runs",lambda x: (x == 4).sum()),
        sixes=("batsman_runs",lambda x: (x == 6).sum()),
        wides=("wide_runs","sum"),
        noballs=("noball_runs","sum"),
        legbyes=("legbye_runs","sum"),
        extras=("extra_runs","sum")
    )
    .reset_index()
    .sort_values("runs", ascending=False)
)

over_analysis.head(10)

,match_id,inning,over,bowler,runs,fours,sixes,wides,noballs,legbyes,extras
2384,32,2,3,Scott Kuggeleijn,32,3,2,4,0,0,4
777,11,1,18,Ruben Trumpelmann,31,1,4,0,1,1,2
497,7,2,9,Rohit Paudel,27,0,4,0,0,0,0
2379,32,2,0,Dipendra Singh Airee,27,0,4,1,0,0,1
384,6,1,12,James Neesham,26,3,2,1,1,0,2
1677,23,1,10,Amar Routela,24,1,2,0,3,0,3
255,4,1,16,Karan KC,24,1,3,0,0,0,0
1873,25,2,12,Basil Hameed,24,4,1,0,0,0,0
262,4,1,19,Pratish GC,23,0,3,1,0,0,1
840,12,1,11,Naren Bhatta,23,0,3,2,0,0,2


### Conclusion: Most Expensive Overs in the Nepal Premier League

The analysis of the highest runs conceded in a single over highlights some of the most expensive overs in the Nepal Premier League. These overs had a significant impact on match momentum, often changing the course of the innings and providing batting teams with crucial scoring opportunities.

Scott Kuggeleijn conceded the most runs in a single over, allowing **32 runs** in an over that included **three fours, two sixes, and four wides**. This combination of boundaries and extras demonstrates how a single over can become extremely costly when both attacking batting and bowling inaccuracies occur simultaneously.

Ruben Trumpelmann conceded **31 runs** in an over that contained **four sixes**, emphasizing the influence of power hitting in the T20 format. Similarly, Rohit Paudel and Dipendra Singh Airee each conceded **27 runs**, largely due to multiple sixes within the over.

Several important patterns emerge from this analysis:

* Sixes were the primary contributors to high-scoring overs, with many expensive overs containing three or four maximums.
* Extras such as wides and no-balls significantly increased the total runs conceded, putting additional pressure on bowlers.
* Both specialist bowlers and all-rounders appear in this list, indicating that any bowler can experience difficult overs in the T20 format.
* Expensive overs often occur during aggressive batting phases, particularly the middle and death overs, where batters actively target bowlers.

The analysis also shows that bowling discipline is crucial in T20 cricket. Overs containing wides and no-balls not only increase the score but also provide additional deliveries to the batting side, allowing further scoring opportunities.

Overall, these expensive overs illustrate the high-risk and high-reward nature of T20 cricket. While a single costly over does not define a bowler's overall ability, understanding these situations helps teams evaluate pressure handling, bowling strategies, and match situations where bowlers may require additional support or tactical adjustments.


# Question No 7 : Best Economy And Worst Economy  on NPL 

In [ ]:
WITH bowler_stats AS (
    SELECT
        bowler,

        COUNT(*) AS balls,

        ROUND(COUNT(*) / 6, 1) AS overs,

        SUM(total_runs) AS runs,

        SUM(wide + noball) AS extras,

        ROUND(
            SUM(total_runs) /
            NULLIF(COUNT(*) / 6, 0),
        2) AS economy

    FROM npl_ball_by_ball

    GROUP BY bowler

    HAVING COUNT(*) >= 300
)

SELECT
    bowler,
    balls,
    overs,
    runs,
    extras,
    economy

FROM bowler_stats

ORDER BY economy ASC;

## Worst Economy 

In [ ]:
WITH bowler_stats AS (
    SELECT
        bowler,

        COUNT(*) AS balls,

        ROUND(COUNT(*) / 6, 1) AS overs,

        SUM(total_runs) AS runs,

        SUM(wide + noball) AS extras,

        ROUND(
            SUM(total_runs) /
            NULLIF(COUNT(*) / 6, 0),
        2) AS economy

    FROM npl_ball_by_ball

    GROUP BY bowler

    HAVING COUNT(*) >= 300
)

SELECT
    bowler,
    balls,
    overs,
    runs,
    extras,
    economy

FROM bowler_stats

ORDER BY economy DESC;

In [51]:
final.columns

Index(['season', 'match_id', 'inning', 'batting_team', 'bowling_team',
       'ball_over', 'ball_result', 'bowler', 'batsman', 'non_striker',
       'shot_direction', 'player_dismissed', 'dismissal_kind', 'fielder',
       'batsman_runs', 'wide_runs', 'bye_runs', 'legbye_runs', 'noball_runs',
       'extra_runs', 'total_runs', 'fielders', 'over'],
      dtype='object')

In [54]:
bowler_stats = (
    final.groupby("bowler")
      .agg(
          balls=("bowler", "size"),
          runs=("total_runs", "sum"),
          extras=("wide_runs", "sum")
      )
      .reset_index()
)

bowler_stats["extras"] += (
    final.groupby("bowler")["noball_runs"]
      .sum()
      .values
)

bowler_stats["overs"] = (
    bowler_stats["balls"] / 6
)

bowler_stats["economy"] = (
    bowler_stats["runs"] /
    bowler_stats["overs"]
).round(2)

bowler_stats = bowler_stats[
    bowler_stats["balls"] >= 300
]

bowler_stats 

,bowler,balls,runs,extras,overs,economy
6,Abinash Bohara,302,380,15,50.333333,7.55
46,Harmeet Singh,326,297,15,54.333333,5.47
57,Karan KC,331,384,15,55.166667,6.96
63,Lalit Rajbanshi,360,358,12,60.000000,5.97
80,Nandan Yadav,307,408,24,51.166667,7.97
109,Sandeep Lamichhane,389,375,5,64.833333,5.78
114,Scott Kuggeleijn,353,371,44,58.833333,6.31
115,Shahab Alam,308,301,2,51.333333,5.86
118,Sher Malla,336,368,3,56.000000,6.57
120,Sohail Tanvir,307,316,14,51.166667,6.18


# Best Economy 

In [55]:
best_economy = (
    bowler_stats
    .sort_values("economy")
)

best_economy.head(10)

,bowler,balls,runs,extras,overs,economy
46,Harmeet Singh,326,297,15,54.333333,5.47
109,Sandeep Lamichhane,389,375,5,64.833333,5.78
115,Shahab Alam,308,301,2,51.333333,5.86
63,Lalit Rajbanshi,360,358,12,60.000000,5.97
120,Sohail Tanvir,307,316,14,51.166667,6.18
114,Scott Kuggeleijn,353,371,44,58.833333,6.31
118,Sher Malla,336,368,3,56.000000,6.57
57,Karan KC,331,384,15,55.166667,6.96
6,Abinash Bohara,302,380,15,50.333333,7.55
80,Nandan Yadav,307,408,24,51.166667,7.97


### Conclusion: Best Economy Bowlers in the Nepal Premier League

The economy rate analysis provides valuable insights into the bowlers who have been most effective at restricting runs in the Nepal Premier League. Economy rate is one of the most important bowling metrics in T20 cricket, as it measures a bowler's ability to control scoring opportunities and maintain pressure on the batting side.

Harmeet Singh emerged as the most economical bowler, recording an outstanding economy rate of **5.47 runs per over** across more than 54 overs. This performance demonstrates exceptional consistency, accuracy, and the ability to limit scoring opportunities throughout the tournament.

Sandeep Lamichhane, who bowled the highest number of deliveries among all bowlers, maintained an impressive economy of **5.78** over nearly 65 overs. His ability to combine a heavy workload with excellent run control highlights his importance to the team's bowling attack.

Shahab Alam and Lalit Rajbanshi also delivered highly economical performances, with economy rates below six runs per over. These figures indicate their effectiveness in building pressure, controlling the middle overs, and forcing batters into mistakes.

Several important observations can be drawn from this analysis:

* Bowlers with lower economy rates consistently restricted scoring opportunities.
* Spin bowlers appear to have played a major role in controlling the run rate during the tournament.
* High workloads combined with low economy rates indicate both consistency and reliability.
* Extras conceded have a noticeable impact on economy rates, emphasizing the importance of bowling discipline.

While bowlers such as Karan KC, Abinash Bohara, and Nandan Yadav bowled significant numbers of overs, their higher economy rates suggest that they conceded runs more frequently. However, these bowlers may also have performed more aggressive roles such as bowling during powerplay or death overs, where scoring rates are generally higher.

Overall, the economy analysis identifies Harmeet Singh, Sandeep Lamichhane, Shahab Alam, and Lalit Rajbanshi as the most effective run-controlling bowlers in the Nepal Premier League. Their ability to maintain low economy rates while bowling substantial workloads makes them valuable assets and key contributors to their teams' bowling strategies.


# Worst Economy 

In [56]:
worst_economy = (
    bowler_stats
    .sort_values("economy", ascending=False)
)

worst_economy.head(10)

,bowler,balls,runs,extras,overs,economy
121,Sompal Kami,341,465,23,56.833333,8.18
80,Nandan Yadav,307,408,24,51.166667,7.97
6,Abinash Bohara,302,380,15,50.333333,7.55
57,Karan KC,331,384,15,55.166667,6.96
118,Sher Malla,336,368,3,56.000000,6.57
114,Scott Kuggeleijn,353,371,44,58.833333,6.31
120,Sohail Tanvir,307,316,14,51.166667,6.18
63,Lalit Rajbanshi,360,358,12,60.000000,5.97
115,Shahab Alam,308,301,2,51.333333,5.86
109,Sandeep Lamichhane,389,375,5,64.833333,5.78


### Conclusion: Worst Economy Bowlers in the Nepal Premier League

The economy rate analysis of the least economical bowlers highlights the challenges faced by certain bowlers in controlling the scoring rate during the Nepal Premier League. In T20 cricket, bowlers who concede runs at a higher rate often operate in high-pressure situations such as the powerplay and death overs, where batters tend to play aggressively.

Sompal Kami recorded the highest economy rate among the analyzed bowlers, conceding **8.18 runs per over** across nearly 57 overs. Despite his substantial bowling workload, the higher economy rate suggests that opposition batters scored freely against him during several matches.

Nandan Yadav and Abinash Bohara also experienced relatively expensive tournaments, with economy rates of **7.97** and **7.55**, respectively. Both bowlers bowled more than 50 overs, indicating that they were regularly trusted by their teams despite conceding runs.

Several important insights emerge from this analysis:

* Bowlers with higher economy rates often bowl in difficult match situations, including powerplay and death overs.
* A large workload combined with a higher economy rate indicates that captains continued to rely on these bowlers despite challenging conditions.
* Extras contributed to increased economy rates, particularly for bowlers who conceded wides and no-balls.
* Fast bowlers generally face greater risk in T20 cricket due to aggressive batting strategies.

Interestingly, some bowlers appearing in this analysis, such as Karan KC, Sher Malla, and Scott Kuggeleijn, are also among the tournament's leading wicket-takers. This suggests that these bowlers may adopt an attacking approach, prioritizing wickets even at the expense of conceding additional runs.

In contrast, bowlers such as Sandeep Lamichhane, Shahab Alam, and Lalit Rajbanshi maintained significantly lower economy rates, demonstrating superior control and consistency.

Overall, a higher economy rate should not be evaluated in isolation. Factors such as bowling role, match situation, workload, and wicket-taking ability must also be considered. This analysis helps teams identify areas for improvement, understand bowling risks, and develop more effective bowling strategies for future matches.


 # Question no 9 :  Best Death overs Bowler's in NPL

In [ ]:
WITH death_overs AS (
    SELECT
        bowler,
        match_id,
        FLOOR(over_ball) AS over_no,
        total_runs,
        dismissal_kind

    FROM npl_ball_by_ball

    WHERE FLOOR(over_ball) IN (16,17,18,19,20)
),

bowler_stats AS (
    SELECT
        bowler,

        COUNT(*) / 6 AS overs,

        SUM(total_runs) AS runs,

        ROUND(
            SUM(total_runs) /
            NULLIF(COUNT(*) / 6, 0),
        2) AS economy

    FROM death_overs

    GROUP BY bowler
),

wickets AS (
    SELECT
        bowler,

        COUNT(*) AS wickets

    FROM death_overs

    WHERE LOWER(dismissal_kind) IN (
        'caught',
        'bowled',
        'lbw',
        'stumped',
        'caught & bowled',
        'hit wicket'
    )

    GROUP BY bowler
)

SELECT
    bs.bowler,
    ROUND(bs.overs,1) AS overs,
    bs.runs,

    COALESCE(w.wickets,0) AS wickets,

    bs.economy,

    ROUND(
        COALESCE(w.wickets,0) /
        NULLIF(bs.overs,0),
    2) AS wicket_rate

FROM bowler_stats bs

LEFT JOIN wickets w
    ON bs.bowler = w.bowler

WHERE bs.overs >= 10

ORDER BY
    economy ASC,
    wicket_rate DESC;

In [62]:
death = final[
    final["over"].isin([16,17,18,19,20])
]

In [64]:
bowler_stats = (
    death.groupby("bowler")
    .agg(
        balls=("bowler", "size"),
        runs=("total_runs", "sum")
    )
    .reset_index()
)

bowler_stats["overs"] = (
    bowler_stats["balls"] / 6
)

bowler_stats["economy"] = (
    bowler_stats["runs"] /
    bowler_stats["overs"]
).round(2)

In [65]:
valid_wickets = [
    "caught",
    "bowled",
    "lbw",
    "stumped",
    "caught & bowled",
    "hit wicket"
]

wickets = (
    death[
        death["dismissal_kind"]
        .isin(valid_wickets)
    ]
    .groupby("bowler")
    .size()
    .reset_index(name="wickets")
)

In [66]:
result = bowler_stats.merge(
    wickets,
    on="bowler",
    how="left"
)

result["wickets"] = (
    result["wickets"]
    .fillna(0)
)

result["wicket_rate"] = (
    result["wickets"] /
    result["overs"]
).round(2)

In [67]:
result

,bowler,balls,runs,overs,economy,wickets,wicket_rate
0,Aaditya Mahata,6,5,1.000000,5.00,0.0,0.00
1,Aakash Chand,38,73,6.333333,11.53,3.0,0.47
2,Abhishek Gautam,6,8,1.000000,8.00,1.0,1.00
3,Abhishesh Gautam,34,58,5.666667,10.24,1.0,0.18
4,Abinash Bohara,87,119,14.500000,8.21,10.0,0.69
...,...,...,...,...,...,...,...
83,Surya Tamang,11,27,1.833333,14.73,0.0,0.00
84,Thomas Draca,2,6,0.333333,18.00,0.0,0.00
85,Tilak Bhandari,33,43,5.500000,7.82,3.0,0.55
86,Wayne Parnell,13,14,2.166667,6.46,0.0,0.00


In [68]:
best_death = (
    result[result["overs"] >= 10]
    .sort_values(
        ["economy", "wicket_rate"],
        ascending=[True, False]
    )
)

best_death.head(10)

,bowler,balls,runs,overs,economy,wickets,wicket_rate
70,Sandeep Lamichhane,67,66,11.166667,5.91,8.0,0.72
78,Sohail Tanvir,67,82,11.166667,7.34,9.0,0.81
74,Scott Kuggeleijn,79,97,13.166667,7.37,7.0,0.53
4,Abinash Bohara,87,119,14.500000,8.21,10.0,0.69
36,Karan KC,112,160,18.666667,8.57,11.0,0.59
53,Pratish GC,85,124,14.166667,8.75,10.0,0.71
38,Kishore Mahato,83,126,13.833333,9.11,11.0,0.80
58,Rashid Khan,74,115,12.333333,9.32,6.0,0.49
79,Sompal Kami,124,196,20.666667,9.48,9.0,0.44
61,Rijan Dhakal,98,156,16.333333,9.55,4.0,0.24


## Conclusion: Best Death Overs Bowlers in the Nepal Premier League

The death overs analysis (overs 16–20) evaluates bowlers based on their ability to control runs and take wickets during the most challenging phase of a T20 innings. Since batters generally attack aggressively in the final overs, bowlers who perform well during this period are considered highly valuable to their teams.

Sandeep Lamichhane emerged as the most economical death-over bowler, conceding only 5.91 runs per over while also taking 8 wickets at an impressive wicket rate of 0.72 wickets per over. His ability to both restrict scoring and take wickets demonstrates exceptional control under pressure.

Sohail Tanvir also delivered outstanding performances during the death overs. Despite bowling in aggressive situations, he maintained an economy rate of 7.34 while taking 9 wickets, resulting in the highest wicket rate among the leading bowlers. This indicates his effectiveness as a strike bowler during crucial stages of the innings.

Scott Kuggeleijn and Abinash Bohara also contributed significantly, combining wicket-taking ability with relatively controlled economy rates. Karan KC and Pratish GC further demonstrated their effectiveness by taking double-digit wickets during the death overs.

Several important insights emerge from this analysis:

Lower economy rates during death overs indicate strong control and effective execution under pressure.
Higher wicket rates suggest the ability to dismiss batters during aggressive scoring phases.
Bowlers who can both restrict runs and take wickets provide the greatest value to their teams.
Death-over specialists play a crucial role in limiting opposition totals and defending targets.

Although bowlers such as Sompal Kami and Rijan Dhakal recorded higher economy rates, they still bowled significant numbers of overs, indicating the trust placed in them during difficult match situations.

Overall, Sandeep Lamichhane and Sohail Tanvir stand out as the most effective death-over bowlers in the tournament. Their combination of economy and wicket-taking ability makes them key assets during the final overs and demonstrates their importance in high-pressure situations

In [69]:
worst_death = (
    result[result["overs"] >= 10]
    .sort_values(
        ["economy", "wicket_rate"],
        ascending=[False, True]
    )
)

worst_death.head(10)

,bowler,balls,runs,overs,economy,wickets,wicket_rate
33,James Neesham,68,124,11.333333,10.94,5.0,0.44
50,Nandan Yadav,123,210,20.500000,10.24,12.0,0.59
64,Ruben Trumpelmann,63,106,10.500000,10.10,4.0,0.38
44,Marchant de Lange,72,121,12.000000,10.08,3.0,0.25
61,Rijan Dhakal,98,156,16.333333,9.55,4.0,0.24
79,Sompal Kami,124,196,20.666667,9.48,9.0,0.44
58,Rashid Khan,74,115,12.333333,9.32,6.0,0.49
38,Kishore Mahato,83,126,13.833333,9.11,11.0,0.80
53,Pratish GC,85,124,14.166667,8.75,10.0,0.71
36,Karan KC,112,160,18.666667,8.57,11.0,0.59


### Conclusion: Worst Death Overs Bowlers in the Nepal Premier League

The death overs (overs 16–20) represent the most demanding phase of a T20 innings, where batters aggressively target bowlers to maximize scoring opportunities. The analysis of the least economical death-over bowlers identifies those who faced challenges in controlling runs during these crucial stages.

James Neesham recorded the highest economy rate among the analyzed bowlers, conceding **10.94 runs per over** during the death overs. Despite taking five wickets, the high scoring rate against him indicates difficulties in restricting batters during the final overs.

Nandan Yadav also experienced a challenging death-over performance, conceding **210 runs** at an economy rate of **10.24**. However, his **12 wickets** demonstrate that he adopted an aggressive wicket-taking approach, often sacrificing economy in an attempt to dismiss batters.

Ruben Trumpelmann and Marchant de Lange similarly struggled to contain scoring, with economy rates exceeding ten runs per over. Their lower wicket rates suggest that they found it difficult to both control runs and take wickets consistently during this phase.

Several important observations can be drawn from this analysis:

* Death overs naturally produce higher scoring rates because batters play aggressively.
* Bowlers assigned to the final overs often face the strongest hitters under maximum pressure.
* High economy rates do not necessarily indicate poor bowling ability, as some bowlers deliberately attack in search of wickets.
* Wicket-taking bowlers may concede additional runs while attempting to break partnerships or prevent bigger totals.

Interestingly, bowlers such as Nandan Yadav, Karan KC, and Kishore Mahato recorded relatively high wicket counts despite their higher economy rates. This suggests that they played attacking roles and were trusted by their teams during crucial moments.

Overall, the analysis highlights the difficult nature of death-over bowling in T20 cricket. While these bowlers conceded runs at higher rates, many continued to contribute through wicket-taking. Therefore, death-over performance should be evaluated by balancing both economy and wicket-taking ability rather than relying solely on runs conceded.


# Question 10 : Best and Worst  powerplay overs Bowler's in NPL

In [ ]:
WITH powerplay_data AS (
    SELECT
        bowler,
        match_id,
        FLOOR(over_ball) AS over_no,
        total_runs,
        LOWER(TRIM(dismissal_kind)) AS dismissal_kind

    FROM npl_ball_by_ball

    WHERE FLOOR(over_ball) BETWEEN 1 AND 6
),

bowler_stats AS (
    SELECT
        bowler,

        COUNT(*) / 6 AS overs,

        SUM(total_runs) AS runs,

        ROUND(
            SUM(total_runs) /
            NULLIF(COUNT(*) / 6, 0),
        2) AS economy

    FROM powerplay_data

    GROUP BY bowler
),

powerplay_wickets AS (
    SELECT
        bowler,

        COUNT(*) AS wickets

    FROM powerplay_data

    WHERE dismissal_kind IN (
        'caught',
        'bowled',
        'lbw',
        'stumped',
        'caught & bowled',
        'hit wicket'
    )

    GROUP BY bowler
)

SELECT
    bs.bowler,
    ROUND(bs.overs,1) AS overs,
    bs.runs,

    COALESCE(w.wickets,0) AS wickets,

    bs.economy,

    ROUND(
        COALESCE(w.wickets,0) /
        NULLIF(bs.overs,0),
    2) AS wicket_rate

FROM bowler_stats bs

LEFT JOIN powerplay_wickets w
    ON bs.bowler = w.bowler

WHERE bs.overs >= 10

ORDER BY
    economy ASC,
    wicket_rate DESC;

# poopr powerplay 

In [70]:
powerplay = final[
    final["over"].between(1, 6)
]

In [71]:
bowler_stats = (
    powerplay.groupby("bowler")
    .agg(
        balls=("bowler", "size"),
        runs=("total_runs", "sum")
    )
    .reset_index()
)

bowler_stats["overs"] = (
    bowler_stats["balls"] / 6
)

bowler_stats["economy"] = (
    bowler_stats["runs"] /
    bowler_stats["overs"]
).round(2)

In [72]:
valid_wickets = [
    "caught",
    "bowled",
    "lbw",
    "stumped",
    "caught & bowled",
    "hit wicket"
]

wickets = (
    powerplay[
        powerplay["dismissal_kind"]
        .isin(valid_wickets)
    ]
    .groupby("bowler")
    .size()
    .reset_index(name="wickets")
)

In [73]:
result = bowler_stats.merge(
    wickets,
    on="bowler",
    how="left"
)

result["wickets"] = (
    result["wickets"]
    .fillna(0)
)

result["wicket_rate"] = (
    result["wickets"] /
    result["overs"]
).round(2)

In [74]:
best_powerplay = (
    result[result["overs"] >= 10]
    .sort_values(
        ["economy", "wicket_rate"],
        ascending=[True, False]
    )
)

best_powerplay.head(10)

,bowler,balls,runs,overs,economy,wickets,wicket_rate
76,Saif Zaib,90,70,15.000000,4.67,5.0,0.33
98,William Bosisto,74,60,12.333333,4.86,4.0,0.32
88,Sohail Tanvir,126,106,21.000000,5.05,8.0,0.38
54,Milind Kumar,102,88,17.000000,5.18,8.0,0.47
56,Nandan Yadav,93,81,15.500000,5.23,5.0,0.32
46,Lalit Rajbanshi,178,156,29.666667,5.26,10.0,0.34
99,Zeeshan Maqsood,75,66,12.500000,5.28,4.0,0.32
71,Ruben Trumpelmann,85,75,14.166667,5.29,6.0,0.42
11,Basir Ahamad,79,71,13.166667,5.39,4.0,0.30
32,Harsh Thaker,60,55,10.000000,5.50,1.0,0.10


### Conclusion: Best Powerplay Bowlers in the Nepal Premier League

The powerplay phase (overs 1–6) is one of the most critical periods in T20 cricket, as only two fielders are allowed outside the 30-yard circle. Bowlers who perform well during these overs play a crucial role in restricting scoring opportunities and providing early breakthroughs.

Saif Zaib emerged as the most economical powerplay bowler, conceding only **4.67 runs per over** across 15 overs. His ability to maintain such a low economy rate during the field restriction period demonstrates exceptional control, accuracy, and discipline.

William Bosisto and Sohail Tanvir also delivered highly effective performances, recording economy rates below **5.10 runs per over**. Sohail Tanvir, in particular, combined economical bowling with wicket-taking ability, claiming **8 wickets** during the powerplay phase.

Milind Kumar proved to be both economical and effective, maintaining an economy of **5.18** while achieving a strong wicket rate of **0.47 wickets per over**. Ruben Trumpelmann also performed well by balancing run control with wicket-taking, recording a wicket rate of **0.42**.

Several important observations emerge from this analysis:

* Economy rates below six runs per over during the powerplay indicate excellent bowling performance.
* Bowlers who can both restrict runs and take wickets provide maximum value to their teams.
* Early wickets can significantly impact the opposition's scoring potential throughout the innings.
* Spin bowlers and medium pacers appear to have adapted effectively to powerplay conditions.

Lalit Rajbanshi deserves special recognition for maintaining an economy of **5.26** while bowling nearly **30 overs** during the powerplay, indicating both consistency and the trust placed in him by his team.

Overall, Saif Zaib, Sohail Tanvir, Milind Kumar, and Lalit Rajbanshi emerged as the most effective powerplay bowlers in the tournament. Their ability to control scoring rates and take early wickets provided their teams with a significant advantage and established strong foundations for the remainder of the innings.


In [75]:
poor_powerplay = (
    result[result["overs"] >= 10]
    .sort_values(
        ["economy", "wicket_rate"],
        ascending=[False, True]
    )
)

poor_powerplay.head(10)

,bowler,balls,runs,overs,economy,wickets,wicket_rate
66,Rashid Khan,61,94,10.166667,9.25,2.0,0.20
89,Sompal Kami,100,144,16.666667,8.64,5.0,0.30
25,Dipendra Singh Airee,115,161,19.166667,8.40,9.0,0.47
33,Hemant Dhami,63,84,10.500000,8.00,6.0,0.57
19,Chris Sole,73,94,12.166667,7.73,3.0,0.25
43,Kishore Mahato,81,104,13.500000,7.70,1.0,0.07
86,Sher Malla,108,138,18.000000,7.67,9.0,0.50
62,Pratish GC,134,171,22.333333,7.66,4.0,0.18
67,Raymon Reifer,63,79,10.500000,7.52,2.0,0.19
38,James Neesham,60,75,10.000000,7.50,2.0,0.20


### Conclusion: Poor Powerplay Bowlers in the Nepal Premier League

The powerplay overs (overs 1–6) are among the most challenging phases for bowlers in T20 cricket due to fielding restrictions and aggressive batting approaches. The analysis of the least economical powerplay bowlers highlights those who struggled to contain runs during the early stages of the innings.

Rashid Khan recorded the highest economy rate in the powerplay, conceding **9.25 runs per over** while taking only **2 wickets**. Although widely recognized for his bowling abilities, these figures suggest that he faced difficulties in controlling runs during the opening overs of the tournament.

Sompal Kami and Dipendra Singh Airee also experienced relatively expensive powerplay spells, with economy rates of **8.64** and **8.40**, respectively. However, Dipendra Singh Airee managed to compensate through wicket-taking, claiming **9 wickets** and achieving a strong wicket rate of **0.47 wickets per over**.

Hemant Dhami represents an interesting case, conceding **8.00 runs per over** while maintaining an excellent wicket rate of **0.57**, one of the highest among all powerplay bowlers. This indicates an attacking approach where wickets were prioritized over economy.

Several key observations emerge from this analysis:

* Higher economy rates are expected during the powerplay due to field restrictions.
* Some bowlers sacrificed economy to attack aggressively and take early wickets.
* Wicket-taking ability can offset higher run rates, particularly in T20 cricket.
* Bowlers with both high economy rates and low wicket rates may require tactical adjustments during the powerplay.

Kishore Mahato and Pratish GC experienced challenges in both economy and wicket-taking efficiency, indicating difficulties in controlling scoring opportunities while also struggling to take regular wickets.

Interestingly, several bowlers appearing in this list, including Sompal Kami, Sher Malla, and Dipendra Singh Airee, remained trusted by their teams and bowled substantial numbers of powerplay overs. This suggests that captains valued their experience and ability to create breakthroughs despite conceding runs.

Overall, the analysis demonstrates that powerplay bowling involves balancing two competing objectives: restricting runs and taking early wickets. While these bowlers conceded runs at higher rates, some continued to provide value through their wicket-taking abilities, emphasizing the importance of evaluating both economy and wicket rate together rather than considering economy alone.


# Question no 11 : Best and Worst Middle overs Bowler's in Nepal Premier League

In [ ]:
WITH middle_over_data AS (
    SELECT
        bowler,
        match_id,
        FLOOR(over_ball) AS over_no,
        total_runs,
        LOWER(TRIM(dismissal_kind)) AS dismissal_kind

    FROM npl_ball_by_ball

    WHERE FLOOR(over_ball) BETWEEN 7 AND 15
),

bowler_stats AS (
    SELECT
        bowler,

        COUNT(*) AS balls,

        ROUND(COUNT(*) / 6, 1) AS overs,

        SUM(total_runs) AS runs,

        ROUND(
            SUM(total_runs) /
            NULLIF(COUNT(*) / 6, 0),
        2) AS economy

    FROM middle_over_data

    GROUP BY bowler
),

middle_wickets AS (
    SELECT
        bowler,

        COUNT(*) AS wickets

    FROM middle_over_data

    WHERE dismissal_kind IN (
        'caught',
        'bowled',
        'lbw',
        'stumped',
        'caught & bowled',
        'hit wicket'
    )

    GROUP BY bowler
)

SELECT
    bs.bowler,
    bs.overs,
    bs.runs,

    COALESCE(w.wickets, 0) AS wickets,

    bs.economy,

    ROUND(
        COALESCE(w.wickets,0) /
        NULLIF(bs.overs,0),
    2) AS wicket_rate

FROM bowler_stats bs

LEFT JOIN middle_wickets w
    ON bs.bowler = w.bowler

WHERE bs.overs >= 10

ORDER BY
    economy ASC,
    wicket_rate DESC;

# poor 

In [77]:
middle = final[
    final["over"].between(7, 15)
]

In [78]:
bowler_stats = (
    middle.groupby("bowler")
    .agg(
        balls=("bowler", "size"),
        runs=("total_runs", "sum")
    )
    .reset_index()
)

bowler_stats["overs"] = (
    bowler_stats["balls"] / 6
)

bowler_stats["economy"] = (
    bowler_stats["runs"] /
    bowler_stats["overs"]
).round(2)

In [79]:
valid_wickets = [
    "caught",
    "bowled",
    "lbw",
    "stumped",
    "caught & bowled",
    "hit wicket"
]

wickets = (
    middle[
        middle["dismissal_kind"]
        .isin(valid_wickets)
    ]
    .groupby("bowler")
    .size()
    .reset_index(name="wickets")
)

In [80]:
result = bowler_stats.merge(
    wickets,
    on="bowler",
    how="left"
)

result["wickets"] = (
    result["wickets"]
    .fillna(0)
)

result["wicket_rate"] = (
    result["wickets"] /
    result["overs"]
).round(2)

In [81]:
best_middle = (
    result[result["overs"] >= 10]
    .sort_values(
        ["economy", "wicket_rate"],
        ascending=[True, False]
    )
)

best_middle.head(10)

,bowler,balls,runs,overs,economy,wickets,wicket_rate
35,Gerhard Erasmus,154,129,25.666667,5.03,10.0,0.39
119,Tilak Bhandari,121,105,20.166667,5.21,4.0,0.20
124,Zeeshan Maqsood,122,110,20.333333,5.41,6.0,0.30
55,Lalit Rajbanshi,122,113,20.333333,5.56,6.0,0.30
38,Harmeet Singh,142,133,23.666667,5.62,9.0,0.38
105,Shahab Alam,158,148,26.333333,5.62,8.0,0.30
100,Sandeep Lamichhane,244,230,40.666667,5.66,12.0,0.30
39,Harsh Thaker,92,87,15.333333,5.67,6.0,0.39
30,Dipesh Kandel,66,63,11.000000,5.73,5.0,0.45
83,Ranjeet Kumar,66,63,11.000000,5.73,2.0,0.18


### Conclusion: Best Middle Overs Bowlers in the Nepal Premier League

The middle overs phase (overs 7–15) plays a crucial role in T20 cricket, as teams aim to balance scoring acceleration with wicket preservation. Bowlers who perform effectively during this period help control the run rate, build pressure, and create opportunities for wickets.

Gerhard Erasmus emerged as the most effective middle-overs bowler, recording an excellent economy rate of **5.03 runs per over** while taking **10 wickets**. His ability to combine run control with regular breakthroughs makes him one of the most impactful bowlers during this phase.

Tilak Bhandari and Zeeshan Maqsood also delivered highly economical spells, maintaining economy rates close to **five runs per over**. Their performances indicate strong control, disciplined bowling, and the ability to restrict scoring opportunities during the middle stages of the innings.

Lalit Rajbanshi, Harmeet Singh, and Shahab Alam further demonstrated their effectiveness by maintaining economy rates below **5.65**, highlighting the important contribution of spin bowlers in controlling the game during the middle overs.

Several important observations emerge from this analysis:

* Bowlers with economy rates below six runs per over were highly successful in limiting scoring opportunities.
* Middle-over specialists often rely on accuracy, variation, and tactical bowling rather than pace alone.
* Regular wicket-taking during this phase disrupts partnerships and prevents batting teams from accelerating later in the innings.
* Spin bowlers played a significant role in controlling the middle overs.

Sandeep Lamichhane deserves special recognition for bowling more than **40 overs** during the middle phase while maintaining an economy of **5.66** and taking **12 wickets**, demonstrating both consistency and durability.

Dipesh Kandel achieved one of the highest wicket rates (**0.45 wickets per over**), indicating his effectiveness in breaking partnerships despite a relatively smaller bowling workload.

Overall, Gerhard Erasmus, Sandeep Lamichhane, Harmeet Singh, and Lalit Rajbanshi emerged as the leading middle-over bowlers in the tournament. Their ability to restrict runs while taking important wickets provided their teams with crucial control during the most strategic phase of the innings. These bowlers can be considered the primary middle-over specialists of the Nepal Premier League.


# poor Middle overs 

In [82]:
poor_middle = (
    result[result["overs"] >= 10]
    .sort_values(
        ["economy", "wicket_rate"],
        ascending=[False, True]
    )
)

poor_middle.head(10)

,bowler,balls,runs,overs,economy,wickets,wicket_rate
37,Gulshan Jha,64,98,10.666667,9.19,3.0,0.28
16,Bibek Yadav,70,107,11.666667,9.17,2.0,0.17
123,Yuvraj Khatri,66,95,11.000000,8.64,3.0,0.27
121,Unish Singh,60,86,10.000000,8.60,1.0,0.10
45,James Neesham,101,143,16.833333,8.50,5.0,0.30
42,Ishan Pandey,61,84,10.166667,8.26,4.0,0.39
96,Sagar Dhakal,120,161,20.000000,8.05,6.0,0.30
22,D'Arcy Short,76,101,12.666667,7.97,0.0,0.00
6,Amar Routela,70,90,11.666667,7.71,5.0,0.43
122,William Bosisto,146,183,24.333333,7.52,8.0,0.33


### Conclusion: Poor Middle Overs Bowlers in the Nepal Premier League

The middle overs (overs 7–15) are considered the control phase of a T20 innings, where bowling teams aim to restrict scoring and build pressure through disciplined bowling. The analysis of the least effective middle-over bowlers highlights those who struggled to contain runs or take wickets consistently during this critical period.

Gulshan Jha recorded the highest economy rate among the analyzed bowlers, conceding **9.19 runs per over** while taking only **3 wickets**. Similarly, Bibek Yadav and Yuvraj Khatri also experienced difficulties controlling the scoring rate, with economy rates above **8.50 runs per over**.

Unish Singh had one of the lowest wicket rates, claiming only **one wicket** while conceding **8.60 runs per over**. This suggests challenges in both restricting runs and creating wicket-taking opportunities during the middle overs.

James Neesham and Sagar Dhakal bowled a substantial number of overs during this phase but recorded relatively high economy rates. Despite this, both bowlers contributed through wicket-taking, indicating that they were often used in attacking roles.

Several important insights emerge from this analysis:

* High economy rates during the middle overs allow batting teams to maintain momentum and build strong partnerships.
* Lower wicket rates reduce the bowling side's ability to create pressure and slow down scoring.
* Some bowlers struggled in both economy and wicket-taking effectiveness, making them less impactful during this phase.
* Bowlers who concede runs but continue taking wickets may still provide tactical value.

D'Arcy Short represents a notable case, as he conceded nearly **eight runs per over** without taking a wicket. This indicates that while he contributed overs for his team, he was unable to provide breakthroughs during the middle phase.

William Bosisto and Amar Routela, despite appearing in this analysis, managed to take wickets regularly, suggesting that they adopted a more attacking bowling approach.

Overall, the analysis indicates that effective middle-over bowling requires a balance between controlling the scoring rate and taking wickets. Bowlers with both high economy rates and low wicket rates may need tactical adjustments, while those who continue taking wickets can still contribute positively despite conceding runs. These insights can help teams identify areas for improvement and optimize their middle-over bowling strategies.


# Question 13 : Best Bowlers and Worst Bowlers in a NPL

In [ ]:
WITH bowler_runs AS (
    SELECT
        bowler,

        COUNT(*) AS balls,

        ROUND(COUNT(*) / 6,1) AS overs,

        SUM(total_runs) AS runs,

        ROUND(
            SUM(total_runs) /
            NULLIF(COUNT(*) / 6,0),
        2) AS economy

    FROM npl_ball_by_ball

    GROUP BY bowler

    HAVING COUNT(*) >= 300
),

wickets AS (

    SELECT
        bowler,
        COUNT(*) AS wickets

    FROM npl_ball_by_ball

    WHERE LOWER(TRIM(dismissal_kind)) IN (
        'caught',
        'bowled',
        'lbw',
        'stumped',
        'caught & bowled',
        'hit wicket'
    )

    GROUP BY bowler
),

matches AS (

    SELECT
        bowler,
        COUNT(DISTINCT match_id) AS matches

    FROM npl_ball_by_ball

    GROUP BY bowler
),

best_figures AS (

    SELECT
        bowler,

        MAX(
            CONCAT(
                wickets,
                '-',
                runs
            )
        ) AS best

    FROM (

        SELECT
            match_id,
            bowler,

            COUNT(*) AS wickets,

            SUM(total_runs) AS runs

        FROM npl_ball_by_ball

        WHERE LOWER(TRIM(dismissal_kind)) IN (
            'caught',
            'bowled',
            'lbw',
            'stumped',
            'caught & bowled',
            'hit wicket'
        )

        GROUP BY
            match_id,
            bowler

    ) x

    GROUP BY bowler
)

SELECT
    r.bowler,
    m.matches,
    r.runs,
    w.wickets,
    r.economy,
    b.best

FROM bowler_runs r

JOIN wickets w
    ON r.bowler = w.bowler

JOIN matches m
    ON r.bowler = m.bowler

JOIN best_figures b
    ON r.bowler = b.bowler

ORDER BY
    economy ASC,
    wickets DESC;

# Worst Bowlers 

In [83]:
bowler_runs = (
    final.groupby("bowler")
    .agg(
        balls=("bowler", "size"),
        runs=("total_runs", "sum")
    )
    .reset_index()
)

bowler_runs["overs"] = (
    bowler_runs["balls"] / 6
)

bowler_runs["economy"] = (
    bowler_runs["runs"] /
    bowler_runs["overs"]
).round(2)

bowler_runs = bowler_runs[
    bowler_runs["balls"] >= 300
]


bowler_runs

,bowler,balls,runs,overs,economy
6,Abinash Bohara,302,380,50.333333,7.55
46,Harmeet Singh,326,297,54.333333,5.47
57,Karan KC,331,384,55.166667,6.96
63,Lalit Rajbanshi,360,358,60.000000,5.97
80,Nandan Yadav,307,408,51.166667,7.97
109,Sandeep Lamichhane,389,375,64.833333,5.78
114,Scott Kuggeleijn,353,371,58.833333,6.31
115,Shahab Alam,308,301,51.333333,5.86
118,Sher Malla,336,368,56.000000,6.57
120,Sohail Tanvir,307,316,51.166667,6.18


In [84]:
valid_wickets = [
    "caught",
    "bowled",
    "lbw",
    "stumped",
    "caught & bowled",
    "hit wicket"
]

wickets = (
    final[final["dismissal_kind"]
        .isin(valid_wickets)
    ]
    .groupby("bowler")
    .size()
    .reset_index(name="wickets")
)

In [85]:
matches = (
    final.groupby("bowler")["match_id"]
      .nunique()
      .reset_index(name="matches")
)

In [86]:
best_figures = (
    final[
        final["dismissal_kind"]
        .isin(valid_wickets)
    ]
    .groupby(["match_id", "bowler"])
    .agg(
        wickets=("dismissal_kind", "count"),
        runs=("total_runs", "sum")
    )
    .reset_index()
)

best_figures["best"] = (
    best_figures["wickets"].astype(str)
    + "-"
    + best_figures["runs"].astype(str)
)

best_figures = (
    best_figures
    .sort_values(
        ["bowler", "wickets", "runs"],
        ascending=[True, False, True]
    )
    .drop_duplicates("bowler")
    [["bowler", "best"]]
)
best_figures

,bowler,best
390,Aaditya Mahata,2-0
72,Aakash Chand,2-0
190,Abhishek Gautam,3-0
61,Abhishesh Gautam,2-0
45,Abinash Bohara,4-0
...,...,...
450,Tilak Bhandari,3-0
26,Unish Singh,1-0
247,William Bosisto,6-0
27,Yuvraj Khatri,1-0


In [87]:
result = (
    bowler_runs
    .merge(wickets, on="bowler")
    .merge(matches, on="bowler")
    .merge(best_figures, on="bowler")
)
result

,bowler,balls,runs,overs,economy,wickets,matches,best
0,Abinash Bohara,302,380,50.333333,7.55,24,15,4-0
1,Harmeet Singh,326,297,54.333333,5.47,20,12,3-0
2,Karan KC,331,384,55.166667,6.96,25,14,5-0
3,Lalit Rajbanshi,360,358,60.000000,5.97,21,12,4-0
4,Nandan Yadav,307,408,51.166667,7.97,19,13,3-0
5,Sandeep Lamichhane,389,375,64.833333,5.78,23,15,6-0
6,Scott Kuggeleijn,353,371,58.833333,6.31,27,14,5-0
7,Shahab Alam,308,301,51.333333,5.86,19,14,6-0
8,Sher Malla,336,368,56.000000,6.57,24,13,3-0
9,Sohail Tanvir,307,316,51.166667,6.18,22,10,6-0


### Conclusion: Overall Best Bowlers in the Nepal Premier League

The overall bowling analysis combines multiple performance indicators, including overs bowled, runs conceded, economy rate, wickets taken, matches played, and best bowling figures. This provides a comprehensive evaluation of each bowler's contribution throughout the tournament rather than relying on a single statistic.

Among all bowlers, Sandeep Lamichhane emerged as one of the most complete performers. He bowled the highest number of overs (64.8 overs), maintained an outstanding economy rate of *5.78,* claimed 23 wickets, and recorded excellent best figures of 6 wickets in a match. His ability to combine workload, economy, and wicket-taking makes him one of the most valuable bowlers in the competition.

Scott Kuggeleijn finished as the leading wicket-taker with 27 wickets while maintaining a strong economy rate of 6.31. His best bowling figures of 5 wickets demonstrate his ability to produce match-winning performances.

Harmeet Singh was the most economical bowler in the tournament, conceding only 5.47 runs per over across more than 54 overs. His exceptional control and consistency made him highly effective in restricting opposition scoring.

Shahab Alam and Lalit Rajbanshi also delivered impressive performances, maintaining economy rates below six while contributing important wickets for their teams. Shahab Alam's best figures of 6 wickets in a match further highlight his ability to produce outstanding individual performances.

Several important observations emerge from this analysis:

Sandeep Lamichhane demonstrated the best balance between workload, economy, and wicket-taking.
Scott Kuggeleijn established himself as the tournament's premier strike bowler.
Harmeet Singh proved to be the most economical and disciplined bowler.
Karan KC, Sher Malla, and Sohail Tanvir provided valuable wicket-taking contributions while maintaining competitive economy rates.
Bowlers such as Sompal Kami and Nandan Yadav conceded runs at higher rates but continued to contribute through wickets and extensive bowling workloads.

The analysis shows that successful T20 bowlers are not evaluated solely by wickets or economy rates. Instead, the most valuable bowlers are those who consistently balance run control, wicket-taking ability, workload, and match-winning performances.

Overall, Sandeep Lamichhane, Scott Kuggeleijn, Harmeet Singh, Shahab Alam, and Lalit Rajbanshi emerged as the standout bowlers of the Nepal Premier League, each contributing in different ways to their teams' success. Their performances demonstrate the importance of combining consistency, control, and wicket-taking ability in modern T20 cricket.

# Final Conclusion

The Nepal Premier League Bowling Analytics Project successfully transformed raw ball-by-ball cricket data into meaningful performance insights using MySQL and Python (Pandas). By applying advanced cricket analytics techniques commonly used in professional leagues, this project provided a comprehensive evaluation of bowling performance across multiple dimensions.

The analysis revealed that bowling success in T20 cricket cannot be measured by a single statistic. Instead, a combination of economy rate, wicket-taking ability, workload, match situations, and bowling phases provides a more accurate assessment of a bowler's overall contribution.

Several key findings emerged from the analysis:

* **Sandeep Lamichhane** demonstrated exceptional consistency by bowling the highest number of deliveries while maintaining an excellent economy rate and taking crucial wickets.
* **Scott Kuggeleijn** emerged as the leading wicket-taker, highlighting his effectiveness as a strike bowler.
* **Harmeet Singh** proved to be the most economical bowler, showcasing remarkable control and discipline.
* **Shahab Alam** and **Sohail Tanvir** delivered outstanding individual performances, including multiple five-wicket hauls and best bowling figures.
* Different bowlers excelled in different phases of the game, emphasizing the importance of role specialization in modern T20 cricket.

The phase-wise analysis further identified:

* Powerplay specialists capable of taking early wickets.
* Middle-over bowlers who controlled the scoring rate.
* Death-over specialists who performed under high-pressure situations.

The project also demonstrated that some bowlers with relatively high economy rates still contributed significantly through wicket-taking abilities, indicating that aggressive bowling strategies often involve balancing risk and reward.

From a practical perspective, this analytical framework provides valuable support for:

* Team selection decisions.
* Player performance evaluation.
* Match strategy development.
* Opposition analysis.
* Talent identification.
* Bowling role allocation.
* Player scouting and recruitment.

Furthermore, the project highlights the growing importance of data analytics in modern cricket. By converting large volumes of ball-by-ball data into actionable insights, teams can make evidence-based decisions rather than relying solely on traditional statistics or subjective observations.

Overall, this project successfully establishes a data-driven bowling performance evaluation system for the Nepal Premier League. The analytical methods developed in this study can be extended to batting analysis, team performance evaluation, player rankings, predictive modeling, and future cricket analytics applications.

The findings demonstrate that statistical analysis and cricket knowledge can work together to provide deeper insights into player performance, ultimately contributing to better decision-making and improved team strategies in the Nepal Premier League.


In [89]:
import os
os.getcwd()

'R:\\adarsha\\365\\npl_2\\archive'